File: 프로그래머스/unrated/154540. 무인도 여행/무인도 여행.py
from collections import deque

def solution(maps):
    answer = []
    
    # 델타 상 하 좌 우
    dr = [-1,1,0,0]
    dc = [0,0,-1,1]
    
    visited = [[0]*len(maps[0]) for _ in range(len(maps))]
    
    
    sum_group = []
    
    def bfs(row,col,sum_num):
        queue = deque()
        queue.append([row,col])
        visited[row][col] = 1
        
        while queue:
            r,c = queue.popleft()
            for i in range(len(dr)):
                nr = r + dr[i]
                nc = c + dc[i]
                
                if 0 <= nr < len(maps) and 0 <= nc < len(maps[0]):
                    if maps[nr][nc] != 'X' and visited[nr][nc] == 0:
                        visited[nr][nc] = 1
                        queue.append([nr,nc])
                        sum_num += int(maps[nr][nc])
        
        return sum_num
                        
                        
                
    for r in range(len(maps)):
        for c in range(len(maps[0])):
            if maps[r][c] != 'X' and visited[r][c] == 0:
                sum_num = int(maps[r][c])
                sum_group.append(bfs(r,c,sum_num))
    
    if len(sum_group) == 0:
        return [-1]
    else:
        sum_group.sort()
        return sum_group
    

File: 프로그래머스/unrated/154540. 무인도 여행/무인도 여행.js
function solution(maps) {
    var answer = [];
    
    // 델타 상 하 좌 우
    let dr = [-1,1,0,0]
    let dc = [0,0,-1,1]
    
    let visited = []
    
    for ( let a = 0 ; a < maps.length ; a++){
        let arr = new Array(maps[0].length).fill(0)
        visited.push(arr)
    }

    
    function bfs(r,c) {
        let queue = [r,c]
        let num = Number(maps[r][c])
        visited[r][c] = 1
        
        while (queue.length > 0) {
            let row = queue.shift()
            let col = queue.shift()
            
            for (let i = 0; i < dr.length ; i ++) {
                let nr = row + dr[i]
                let nc = col + dc[i]
                
                if ( 0 <= nr && nr < maps.length && 0 <= nc && nc < maps[0].length ) {
                    if ( maps[nr][nc] !== 'X' && visited[nr][nc] === 0 ) {
                        queue.push(nr)
                        queue.push(nc)
                        num += Number(maps[nr][nc])
                        visited[nr][nc] = 1
                    }
                } 
            }
        }
        return num
    }
    for (let r = 0; r < maps.length ; r++) {
        for (let c = 0; c < maps[0].length ; c ++) {
            if (maps[r][c] !== 'X' && visited[r][c] === 0) {
                answer.push(bfs(r,c))
            }
        }
    }
    
    if (answer.length === 0) {
        answer = [-1]
    } else {
        answer = answer.sort((a,b) => a-b)
    }
    return answer;
}